barrier.go

```diff
@@ -0,0 +1,108 @@
+package sync
+
+import (
+	"context"
+	"sync"
+)
+
+// barrier represents a single barrier with multiple waiters.
+type barrier struct {
+	sync.Mutex
+	count int
+	zcs   []*zeroCounter
+}
+
+// wait waits for the barrier to reach a certain target.
+func (b *barrier) wait(ctx context.Context, target int) error {
+	b.Lock()
+
+	// If we're already over the target, return immediately.
+	if target <= b.count {
+		b.Unlock()
+		return nil
+	}
+
+	// Create a zero counter to wait for target - count elements to signal entry.
+	// It also returns if the context fires.
+	zc := newZeroCounter(ctx, target-b.count)
+
+	// TODO: Or do a simpler way where we just pool every 100ms to see if count has
+	// the correct value or ctx fired. Less code and less complexity.
```

Comment:
Maybe there's a simpler way to do this? Initially, I thought of using `wg.WaitGroup` but the problem is that you can't simply `.Wait` with a context. Thus, I created my own `zeroCounter` whose `.wait` unblocks if the context fires or if the barrier is satisfied.